aten/src/ATen/DeviceAccelerator.cpp

```diff
@@ -0,0 +1,31 @@
+#include <ATen/DeviceAccelerator.h>
+#include <ATen/Context.h>
+
+namespace at {
+
+C10_API std::optional<DeviceType> getAccelerator(bool checked) {
+#define CHECK_NO_CUDA \
+  TORCH_CHECK(!at::hasCUDA(), "Cannot have both CUDA and PrivateUse1");
+
+#define CHECK_NO_PU1 \
+  TORCH_CHECK(!is_privateuse1_backend_registered(), "Cannot have both CUDA and PrivateUse1");
+
+    if (is_privateuse1_backend_registered()) {
```

Comment:
This is and will have to remain a runtime check unfortunately. It should be quite cheap right now though.
If there is a perf issue, we can turn it into a flag that we flip during privateuse1 registration to recover perf.